File: src/main/java/org/veupathdb/lib/blast/field/HitSorting.java
package org.veupathdb.lib.blast.field;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.fasterxml.jackson.databind.JsonNode;
import org.veupathdb.lib.blast.util.JSONConstructor;
import org.veupathdb.lib.blast.util.JSONValue;

public enum HitSorting implements JSONValue
{
  ByExpectValue,
  ByBitScore,
  ByTotalScore,
  ByPercentIdentity,
  ByQueryCoverage,
  ;

  public int getValue() {
    return this.ordinal();
  }

  public static HitSorting fromIntValue(int value) {
    return values()[value];
  }

  @JsonCreator
  public static HitSorting fromJSON(JsonNode js) {
    return fromIntValue(js.intValue());
  }

  @Override
  @JsonValue
  public JsonNode toJSON() {
    return JSONConstructor.newInt(this.ordinal());
  }
}

File: src/main/java/org/veupathdb/lib/blast/field/HitSorting.kt
package org.veupathdb.lib.blast.field

import org.veupathdb.lib.blast.util.JSONValue
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.blast.util.JSONConstructor
import com.fasterxml.jackson.annotation.JsonCreator

enum class HitSorting : JSONValue {
  ByExpectValue, ByBitScore, ByTotalScore, ByPercentIdentity, ByQueryCoverage;

  val value = ordinal

  @JsonValue
  override fun toJSON(): JsonNode {
    return JSONConstructor.newInt(ordinal)
  }

  companion object {
    @JvmStatic
    fun fromIntValue(value: Int): HitSorting {
      return values()[value]
    }

    @JvmStatic
    @JsonCreator
    fun fromJSON(js: JsonNode): HitSorting {
      return fromIntValue(js.intValue())
    }
  }
}